src/lib/timing-src-connector.js

```diff
@@ -0,0 +1,218 @@
+        if (typeof velocity === 'number') {
+            if (velocity === 0) {
+                if ((await player.getPaused()) === false) {
+                    player.pause();
+                }
+            }
+            else if (velocity > 0) {
+                if ((await player.getPaused()) === true) {
+                    await player.play()
+                        .catch(async(err) => {
+                            if (err.name === 'NotAllowedError' && options.autoPlayMuted) {
+                                await player.setMuted(true);
+                                await player.play().catch();
+                            }
+                        });
+                    this.updatePlayer(timingObject, player, options);
+                }
+                if ((await player.getPlaybackRate()) !== velocity) {
+                    player.setPlaybackRate(velocity);
+                }
+            }
+        }
+    }
+
+    /**
+     * Since video players do not play with 100% time precision, we need to closely monitor
+     * our player to be sure it remains in sync with the TimingObject.
+     *
+     * If out of sync, we use the current conditions and the options provided to determine
+     * whether to re-sync via setting currentTime or adjusting the playbackRate
+     *
+     * @param {TimingObject} timingObject
+     * @param {PlayerControls} player
+     * @param {TimingSrcConnectorOptions} options
+     * @return {{cancel: (function(): void)}}
+     */
+    maintainPlaybackPosition(timingObject, player, options) {
+        const { allowedDrift, maxAllowedDrift, minCheckInterval, maxRateAdjustment, maxTimeToCatchUp } = options;
+        const syncInterval = Math.min(maxTimeToCatchUp, Math.max(minCheckInterval, maxAllowedDrift));
+
+        const check = async() => {
+            if ((await player.getPaused()) === true) {
+                return;
+            }
+            const diff = timingObject.query().position - (await player.getCurrentTime());
+            const diffAbs = Math.abs(diff);
+            this.log(`Drift: ${diff}`);
+            if (diffAbs > maxAllowedDrift) {
+                await this.adjustSpeed(player, 0);
+                player.setCurrentTime(timingObject.query().position);
+                this.log('Resync by currentTime');
+            }
+            else if (diffAbs > allowedDrift) {
+                const min = diffAbs / maxTimeToCatchUp;
+                const max = maxRateAdjustment;
+                const adjustment = min < max ? (max - min) / 2 : max;
+                await this.adjustSpeed(player, adjustment * Math.sign(diff));
+                this.log('Resync by playbackRate');
+            }
+        };
+        const interval = setInterval(() => check(), syncInterval * 1000);
+        return { cancel: () => clearInterval(interval) };
+    }
+
+    /**
+     * @param {string} msg
+     */
+    log(msg) {
+        this.logger?.(`TimingSrcConnector: ${msg}`);
+    }
+
+    speedAdjustment = 0;
```

Comment:
It's being used as internal state for the `adjustSpeed` method so I declared it near that method. I think if it were in the constructor it would be a bit confusing as it's not on that same level of abstraction as the other items in the constructer and is merely just a dependency for the `adjustSpeed` method.

Reply:
Okay, that logic makes sense to me.